File: frontend/src/lib/api/rss.ts
import { apiClient } from './client'

// ============================================
// Types
// ============================================

export interface RSSFeedParsingPatterns {
  title?: string
  description?: string
  pubDate?: string
  poster?: string
  background?: string
  logo?: string
  category?: string
  magnet?: string
  magnet_regex?: string
  torrent?: string
  torrent_regex?: string
  size?: string
  size_regex?: string
  seeders?: string
  seeders_regex?: string
  category_regex?: string
  episode_name_parser?: string
  // Regex group numbers
  magnet_regex_group?: number
  torrent_regex_group?: number
  size_regex_group?: number
  seeders_regex_group?: number
  category_regex_group?: number
}

export interface RSSFeedFilters {
  title_filter?: string
  title_exclude_filter?: string
  min_size_mb?: number
  max_size_mb?: number
  min_seeders?: number
  category_filter?: string[]
}

export interface RSSFeedMetrics {
  total_items_found: number
  total_items_processed: number
  total_items_skipped: number
  total_errors: number
  last_scrape_duration?: number
  items_processed_last_run: number
  items_skipped_last_run: number
  errors_last_run: number
  skip_reasons: Record<string, number>
}

export interface CatalogPattern {
  id?: string
  name?: string
  regex: string
  enabled: boolean
  case_sensitive: boolean
  target_catalogs: string[]
}

export interface RSSFeedOwner {
  id: string
  email: string
  username?: string
}

// User RSS Feed (new model with user ownership)
export interface UserRSSFeed {
  id: string
  user_id: string
  name: string
  url: string
  is_active: boolean
  source?: string
  torrent_type: string
  auto_detect_catalog: boolean
  parsing_patterns?: RSSFeedParsingPatterns
  filters?: RSSFeedFilters
  metrics?: RSSFeedMetrics
  catalog_patterns?: CatalogPattern[]
  last_scraped_at?: string
  created_at?: string
  updated_at?: string
  // Present for admin view
  user?: RSSFeedOwner
}

export interface UserRSSFeedCreate {
  name: string
  url: string
  is_active?: boolean
  source?: string
  torrent_type?: string
  auto_detect_catalog?: boolean
  parsing_patterns?: RSSFeedParsingPatterns
  filters?: RSSFeedFilters
  catalog_patterns?: CatalogPattern[]
}

export interface UserRSSFeedUpdate {
  name?: string
  url?: string
  is_active?: boolean
  source?: string
  torrent_type?: string
  auto_detect_catalog?: boolean
  parsing_patterns?: RSSFeedParsingPatterns
  filters?: RSSFeedFilters
  catalog_patterns?: CatalogPattern[]
}

export interface TestFeedResult {
  status: 'success' | 'error'
  message: string
  sample_item?: Record<string, unknown>
  detected_patterns?: Record<string, unknown>
  items_count?: number
  regex_results?: Record<string, unknown>
}

export interface RSSSchedulerStatus {
  crontab: string
  next_run?: string
  enabled: boolean
  last_global_run?: string
}

export interface ScrapeResult {
  status: string
  message: string
  items_processed: number
}

export interface BulkStatusResult {
  status: string
  message: string
  updated_count: number
}

// ============================================
// Legacy types (for backward compatibility)
// ============================================

export interface RSSFeed {
  id: number
  name: string
  url: string
  active: boolean
  auto_detect_catalog: boolean
  source?: string
  torrent_type?: string
  parsing_patterns?: Record<string, unknown>
  filters?: Record<string, unknown>
  catalog_patterns?: Array<Record<string, unknown>>
  last_scraped_at?: string
  created_at?: string
  updated_at?: string
}

export interface RSSFeedCreate {
  name: string
  url: string
  active?: boolean
  auto_detect_catalog?: boolean
  source?: string
  torrent_type?: string
  parsing_patterns?: Record<string, unknown>
  filters?: Record<string, unknown>
  catalog_patterns?: Array<Record<string, unknown>>
}

export interface RSSFeedUpdate {
  name?: string
  url?: string
  active?: boolean
  auto_detect_catalog?: boolean
  source?: string
  torrent_type?: string
  parsing_patterns?: Record<string, unknown>
  filters?: Record<string, unknown>
  catalog_patterns?: Array<Record<string, unknown>>
}

// ============================================
// API Client - New User RSS API (JWT Auth)
// ============================================

const USER_RSS_BASE = '/user-rss'

export const userRssApi = {
  /**
   * List RSS feeds (users see own, admins see all with user info)
   */
  list: async (): Promise<UserRSSFeed[]> => {
    return apiClient.get<UserRSSFeed[]>(`${USER_RSS_BASE}/feeds`)
  },

  /**
   * Get a specific RSS feed
   */
  get: async (feedId: string): Promise<UserRSSFeed> => {
    return apiClient.get<UserRSSFeed>(`${USER_RSS_BASE}/feeds/${feedId}`)
  },

  /**
   * Create a new RSS feed
   */
  create: async (data: UserRSSFeedCreate): Promise<UserRSSFeed> => {
    return apiClient.post<UserRSSFeed>(`${USER_RSS_BASE}/feeds`, data)
  },

  /**
   * Update an RSS feed
   */
  update: async (feedId: string, data: UserRSSFeedUpdate): Promise<UserRSSFeed> => {
    return apiClient.put<UserRSSFeed>(`${USER_RSS_BASE}/feeds/${feedId}`, data)
  },

  /**
   * Delete an RSS feed
   */
  delete: async (feedId: string): Promise<void> => {
    return apiClient.delete(`${USER_RSS_BASE}/feeds/${feedId}`)
  },

  /**
   * Test an existing RSS feed
   */
  testFeed: async (feedId: string): Promise<TestFeedResult> => {
    return apiClient.post<TestFeedResult>(`${USER_RSS_BASE}/feeds/${feedId}/test`)
  },

  /**
   * Test an RSS feed URL before creating
   */
  testUrl: async (url: string, patterns?: Record<string, unknown>): Promise<TestFeedResult> => {
    return apiClient.post<TestFeedResult>(`${USER_RSS_BASE}/feeds/test-url`, { url, patterns })
  },

  /**
   * Scrape a single feed
   */
  scrapeFeed: async (feedId: string): Promise<ScrapeResult> => {
    return apiClient.post<ScrapeResult>(`${USER_RSS_BASE}/feeds/${feedId}/scrape`)
  },

  /**
   * Run all scrapers (admin only)
   */
  runAll: async (): Promise<{ status: string; message: string }> => {
    return apiClient.post<{ status: string; message: string }>(`${USER_RSS_BASE}/feeds/run-all`)
  },

  /**
   * Bulk update feed status
   */
  bulkUpdateStatus: async (feedIds: string[], isActive: boolean): Promise<BulkStatusResult> => {
    return apiClient.post<BulkStatusResult>(`${USER_RSS_BASE}/feeds/bulk-status`, {
      feed_ids: feedIds,
      is_active: isActive,
    })
  },

  /**
   * Get scheduler status
   */
  getSchedulerStatus: async (): Promise<RSSSchedulerStatus> => {
    return apiClient.get<RSSSchedulerStatus>(`${USER_RSS_BASE}/scheduler-status`)
  },
}

// ============================================
// Legacy API Client (for backward compatibility)
// ============================================

export const rssApi = {
  /**
   * Get all RSS feeds
   * @deprecated Use userRssApi.list() instead
   */
  list: async (): Promise<RSSFeed[]> => {
    return apiClient.get<RSSFeed[]>('/rss/feeds')
  },

  /**
   * Get a specific RSS feed
   * @deprecated Use userRssApi.get() instead
   */
  get: async (feedId: number): Promise<RSSFeed> => {
    return apiClient.get<RSSFeed>(`/rss/feeds/${feedId}`)
  },

  /**
   * Create a new RSS feed
   * @deprecated Use userRssApi.create() instead
   */
  create: async (data: RSSFeedCreate): Promise<RSSFeed> => {
    return apiClient.post<RSSFeed>('/rss/feeds', data)
  },

  /**
   * Update an RSS feed
   * @deprecated Use userRssApi.update() instead
   */
  update: async (feedId: number, data: RSSFeedUpdate): Promise<RSSFeed> => {
    return apiClient.put<RSSFeed>(`/rss/feeds/${feedId}`, data)
  },

  /**
   * Delete an RSS feed
   * @deprecated Use userRssApi.delete() instead
   */
  delete: async (feedId: number): Promise<{ detail: string }> => {
    return apiClient.delete<{ detail: string }>(`/rss/feeds/${feedId}`)
  },

  /**
   * Test an RSS feed URL
   * @deprecated Use userRssApi.testUrl() instead
   */
  testFeed: async (url: string, patterns?: Record<string, unknown>): Promise<TestFeedResult> => {
    return apiClient.post<TestFeedResult>('/rss/feeds/test-feed', { url, patterns })
  },

  /**
   * Run the RSS feed scraper manually
   * @deprecated Use userRssApi.runAll() instead
   */
  runScraper: async (): Promise<{ detail: string }> => {
    return apiClient.post<{ detail: string }>('/rss/feeds/run')
  },

  /**
   * Activate or deactivate multiple feeds
   * @deprecated Use userRssApi.bulkUpdateStatus() instead
   */
  bulkUpdateStatus: async (feedIds: number[], activate: boolean): Promise<{ detail: string }> => {
    return apiClient.post<{ detail: string }>('/rss/feeds/activate-deactivate-feeds', {
      feed_ids: feedIds,
      activate,
    })
  },
}

File: python-deprecated/db/schemas/rss.py
"""RSS Feed schemas for scraping configuration."""

from datetime import datetime

from pydantic import BaseModel, Field


class ParsingPattern(BaseModel):
    """Individual parsing pattern."""

    field: str
    path: str
    regex: str | None = None
    regex_group: int | None = 0


class RSSFeedParsingPatterns(BaseModel):
    """RSS feed parsing patterns configuration."""

    title: str | None = "title"
    description: str | None = "description"
    pubDate: str | None = "pubDate"
    poster: str | None = None
    background: str | None = None
    logo: str | None = None
    category: str | None = "category"

    # Advanced patterns with regex support
    magnet: str | None = None
    magnet_regex: str | None = None
    torrent: str | None = None
    torrent_regex: str | None = None
    info_hash: str | None = None  # Direct info hash extraction (e.g., torznab:attr[@name="infohash"]@value)
    info_hash_regex: str | None = None
    size: str | None = None
    size_regex: str | None = None
    seeders: str | None = None
    seeders_regex: str | None = None
    category_regex: str | None = None
    episode_name_parser: str | None = None

    # Regex group numbers
    magnet_regex_group: int = 1
    torrent_regex_group: int = 1
    info_hash_regex_group: int = 1
    size_regex_group: int = 1
    seeders_regex_group: int = 1
    category_regex_group: int = 1


class CatalogPattern(BaseModel):
    """Catalog pattern for auto-detection."""

    name: str
    regex: str
    case_sensitive: bool = False
    enabled: bool = True
    target_catalogs: list[str]


class RSSFeedFilters(BaseModel):
    """RSS feed filtering configuration."""

    title_filter: str | None = None
    title_exclude_filter: str | None = None
    min_size_mb: int | None = None
    max_size_mb: int | None = None
    min_seeders: int | None = None
    category_filter: list[str] | None = None


class RSSFeedMetrics(BaseModel):
    """RSS Feed scraping metrics."""

    total_items_found: int = 0
    total_items_processed: int = 0
    total_items_skipped: int = 0
    total_errors: int = 0
    last_scrape_duration: float | None = None
    items_processed_last_run: int = 0
    items_skipped_last_run: int = 0
    errors_last_run: int = 0
    skip_reasons: dict[str, int] = Field(default_factory=dict)


class RSSFeed(BaseModel):
    """RSS feed configuration (Pydantic schema)."""

    id: str | None = Field(alias="_id")
    name: str
    url: str
    parsing_patterns: RSSFeedParsingPatterns = Field(default_factory=RSSFeedParsingPatterns)
    filters: RSSFeedFilters = Field(default_factory=RSSFeedFilters)
    active: bool = True
    last_scraped: datetime | None = None
    source: str | None = None
    torrent_type: str | None = "public"
    auto_detect_catalog: bool | None = False
    catalog_patterns: list[CatalogPattern] | None = Field(default_factory=list)
    metrics: RSSFeedMetrics | None = Field(default_factory=RSSFeedMetrics)


class RSSFeedCreate(BaseModel):
    """Create a new RSS feed."""

    name: str
    url: str
    auto_detect_catalog: bool | None = False
    catalog_patterns: list[CatalogPattern] | None = Field(default_factory=list)
    parsing_patterns: RSSFeedParsingPatterns = RSSFeedParsingPatterns()
    filters: RSSFeedFilters = RSSFeedFilters()
    active: bool = True
    api_password: str | None = None
    source: str | None = None
    torrent_type: str | None = "public"


class RSSFeedUpdate(BaseModel):
    """Update an existing RSS feed."""

    name: str | None = None
    url: str | None = None
    parsing_patterns: RSSFeedParsingPatterns | None = None
    filters: RSSFeedFilters | None = None
    active: bool | None = None
    auto_detect_catalog: bool | None = None
    catalog_patterns: list[CatalogPattern] | None = None
    source: str | None = None
    torrent_type: str | None = None


class RSSFeedBulkImport(BaseModel):
    """Bulk import RSS feeds."""

    feeds: list[RSSFeedCreate]
    api_password: str


class RSSFeedExamine(BaseModel):
    """Examine an RSS feed URL."""

    url: str
    api_password: str


class RSSFeedCatalogPatternSchema(BaseModel):
    """Schema for RSS feed catalog pattern from SQL model."""

    id: int
    name: str | None = None
    regex: str
    enabled: bool = True
    case_sensitive: bool = False
    target_catalogs: list[str] = Field(default_factory=list)

    model_config = {"from_attributes": True}


class RSSFeedSchema(BaseModel):
    """Schema for RSS feed response from SQL model."""

    id: int
    name: str
    url: str
    parsing_patterns: RSSFeedParsingPatterns | None = None
    filters: RSSFeedFilters | None = None
    active: bool = True
    last_scraped: datetime | None = None
    source: str | None = None
    torrent_type: str = "public"
    auto_detect_catalog: bool = False
    catalog_patterns: list[RSSFeedCatalogPatternSchema] = Field(default_factory=list)
    metrics: RSSFeedMetrics | None = None
    created_at: datetime | None = None
    updated_at: datetime | None = None

    model_config = {"from_attributes": True}

    @classmethod
    def from_sql_model(cls, feed) -> "RSSFeedSchema":
        """Create from SQL model RSSFeed."""
        return cls(
            id=feed.id,
            name=feed.name,
            url=feed.url,
            parsing_patterns=RSSFeedParsingPatterns(**feed.parsing_patterns) if feed.parsing_patterns else None,
            filters=RSSFeedFilters(**feed.filters) if feed.filters else None,
            active=feed.active,
            last_scraped=feed.last_scraped,
            source=feed.source,
            torrent_type=feed.torrent_type,
            auto_detect_catalog=feed.auto_detect_catalog,
            catalog_patterns=[
                RSSFeedCatalogPatternSchema(
                    id=p.id,
                    name=p.name,
                    regex=p.regex,
                    enabled=p.enabled,
                    case_sensitive=p.case_sensitive,
                    target_catalogs=p.target_catalogs,
                )
                for p in (feed.catalog_patterns or [])
            ],
            metrics=RSSFeedMetrics(**feed.metrics) if feed.metrics else None,
            created_at=feed.created_at,
            updated_at=feed.updated_at,
        )


# User RSS Feed Schemas


class UserRSSFeedOwner(BaseModel):
    """Owner information for admin view."""

    id: str
    email: str
    username: str | None = None


class UserRSSFeedCatalogPatternSchema(BaseModel):
    """Schema for user RSS feed catalog pattern."""

    id: str
    name: str | None = None
    regex: str
    enabled: bool = True
    case_sensitive: bool = False
    target_catalogs: list[str] = Field(default_factory=list)

    model_config = {"from_attributes": True}


class UserRSSFeedCreate(BaseModel):
    """Create a new user RSS feed."""

    name: str
    url: str
    is_active: bool = True
    source: str | None = None
    torrent_type: str = "public"
    auto_detect_catalog: bool = False
    parsing_patterns: RSSFeedParsingPatterns | None = None
    filters: RSSFeedFilters | None = None
    catalog_patterns: list[CatalogPattern] | None = Field(default_factory=list)


class UserRSSFeedUpdate(BaseModel):
    """Update an existing user RSS feed."""

    name: str | None = None
    url: str | None = None
    is_active: bool | None = None
    source: str | None = None
    torrent_type: str | None = None
    auto_detect_catalog: bool | None = None
    parsing_patterns: RSSFeedParsingPatterns | None = None
    filters: RSSFeedFilters | None = None
    catalog_patterns: list[CatalogPattern] | None = None


class UserRSSFeedResponse(BaseModel):
    """Response schema for user RSS feed."""

    id: str
    user_id: str
    name: str
    url: str
    is_active: bool = True
    source: str | None = None
    torrent_type: str = "public"
    auto_detect_catalog: bool = False
    parsing_patterns: RSSFeedParsingPatterns | None = None
    filters: RSSFeedFilters | None = None
    metrics: RSSFeedMetrics | None = None
    catalog_patterns: list[UserRSSFeedCatalogPatternSchema] = Field(default_factory=list)
    last_scraped_at: datetime | None = None
    created_at: datetime | None = None
    updated_at: datetime | None = None
    user: UserRSSFeedOwner | None = None

    model_config = {"from_attributes": True}

    @classmethod
    def from_sql_model(cls, feed, include_user: bool = False) -> "UserRSSFeedResponse":
        """Create from SQL model UserRSSFeed."""
        response = cls(
            id=feed.id,
            user_id=feed.user_id,
            name=feed.name,
            url=feed.url,
            is_active=feed.is_active,
            source=feed.source,
            torrent_type=feed.torrent_type,
            auto_detect_catalog=feed.auto_detect_catalog,
            parsing_patterns=RSSFeedParsingPatterns(**feed.parsing_patterns) if feed.parsing_patterns else None,
            filters=RSSFeedFilters(**feed.filters) if feed.filters else None,
            metrics=RSSFeedMetrics(**feed.metrics) if feed.metrics else None,
            catalog_patterns=[
                UserRSSFeedCatalogPatternSchema(
                    id=p.id,
                    name=p.name,
                    regex=p.regex,
                    enabled=p.enabled,
                    case_sensitive=p.case_sensitive,
                    target_catalogs=p.target_catalogs,
                )
                for p in (feed.catalog_patterns or [])
            ],
            last_scraped_at=feed.last_scraped_at,
            created_at=feed.created_at,
            updated_at=feed.updated_at,
        )
        if include_user and hasattr(feed, "user") and feed.user:
            response.user = UserRSSFeedOwner(
                id=feed.user.id,
                email=feed.user.email,
                username=feed.user.username,
            )
        return response


class UserRSSFeedTestRequest(BaseModel):
    """Request to test an RSS feed URL."""

    url: str
    patterns: dict | None = None


class UserRSSFeedTestResponse(BaseModel):
    """Response from testing an RSS feed."""

    status: str
    message: str
    sample_item: dict | None = None
    detected_patterns: dict | None = None
    items_count: int | None = None
    regex_results: dict | None = None


class RSSSchedulerStatus(BaseModel):
    """RSS scheduler status information."""

    crontab: str
    next_run: datetime | None = None
    enabled: bool
    last_global_run: datetime | None = None
